Add generic JWT detection and verification

bradlarsen opened the pull request:
### Description:
This adds a generic detector and verifier for generic JWTs.

Detection is simple using regular expressions. However, this will produce many false positives. JWT verification is performed using the `github.com/golang-jwt/jwt/v5` package, performing the usual checks (proper decoding, checking timestamp-related claims, checking for an allow-list of supported algorithms, etc). Only public key cryptography algorithms are supported. Additionally, OIDC Discovery is attempted against the issuer to fetch the public key for signature verification.

### Testing:
A few unit tests demonstrate the detector working. Further testing of verification is needed.

### TODO:
- [x] thoroughly test verification
- [ ] add integration tests
- [x] think about and write up the security implications of using OIDC Discovery with an attacker-controlled URL
- [x] estimate the impact on increased finding volume when verification is disabled

bradlarsen committed:
Fix mixed indentation in alchemy test

bradlarsen committed:
Use `strings.ReplaceAll(...)` instead of `strings.Replace(..., -1)`

bradlarsen committed:
Use integer range loop instead of explicit counting loop

bradlarsen committed:
Revise and expand detectors.go comments

bradlarsen committed:
Draft a generic JWT detector

joeleonjr commented on pkg/detectors/jwt/jwt.go:
```diff
@@ -0,0 +1,215 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch JWKS: %w", err)
+		}
+		defer resp.Body.Close()
```
Haven't used it before but maybe `io.LimitReader()`

bradlarsen committed:
Add comments; simplify JWKS-fetching code

bradlarsen committed:
Fix bugs; refine regex and keywords; refine comments and error messages

bradlarsen committed:
Use github.com/lestrrat-go/jwx to parse JWK sets

bradlarsen committed:
Relax JWT regex to allow for spec-disallowed but sometimes present base64 padding

bradlarsen commented on pkg/detectors/jwt/jwt.go:
```diff
@@ -0,0 +1,215 @@
+			return nil, fmt.Errorf("JWKS returned status: %v", resp.Status)
+		}
+
+		// Load JWKS into `jwt.VerificationKeySet` or `jwt.VerificationKey` from JSON
```
I switched to the https://github.com/lestrrat-go/jwx library for parsing JWK sets to the expected "native" crypto types needed for JWT signature verification.

bradlarsen committed:
Refine error messages; check JWT issuer for localhost and exclude verification

bradlarsen committed:
Update README to note JWT support

bradlarsen committed:
exclude private hosts from JWKS retrieval

bradlarsen committed:
Allow EdDSA JWTs

bradlarsen committed:
Cleanup

bradlarsen committed:
Support unknown verification of HMAC JWTs; include some claims in ExtraData

bradlarsen committed:
cleanup; expand unit tests

bradlarsen commented on pkg/detectors/jwt/jwt_test.go:
```diff
@@ -0,0 +1,83 @@
+package jwt
```
I added a number of additional unit tests. Most of these test cases will in fact need to be integration tests (not yet written) since they are only detected with verification enabled.

bradlarsen commented on pkg/detectors/jwt/jwt.go:
```diff
@@ -0,0 +1,215 @@
+func verifyMatch(ctx context.Context, client *http.Client, tokenString string) (bool, map[string]string, error) {
+
+	// A key retrieval function that uses the OIDC Discovery protocol
+	oidcDiscoveryKeyFunc := func(unverifiedToken *jwt.Token) (any, error) {
```
This is a good idea! I'm going to leave this work as a future extension though.

kashifkhan0771 committed:
Merge branch 'main' into jwt

bradlarsen commented on pkg/detectors/jwt/jwt.go:
```diff
@@ -0,0 +1,312 @@
+// FromData will find and optionally verify JWT secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	client := cmp.Or(s.client, common.SaneHttpClient())
+	seenMatches := make(map[string]bool)
```
I didn't know this! Thinking about it a bit, it makes sense. I'll switch it.

Thanks for the tip.

bradlarsen committed:
Use more efficient set representation

bradlarsen committed:
Do not give HMAC JWTs an `unknown` status

bradlarsen committed:
Include `alg` in extra data

shahzadhaider1 reviewed (dismissed):
Good work. 
Looks good, left one minor comment.

bradlarsen commented on pkg/detectors/jwt/jwt.go:
```diff
@@ -0,0 +1,316 @@
+	case errors.Is(err, jwt.ErrHashUnavailable):
+		return false, err
+	default:
+		return false, nil
```
Yes, that's intentional: the remaining error cases that the `jwt.Parse` function can return (and there are many of them) indicate that the JWT is invalid for some reason, like the timestamp-related claims don't check out.

The code here is not as clear as it might be. I'll at least add a comment. Thanks for pointing this out.

bradlarsen committed:
Be more explicit about JWT verification error propagation

bradlarsen committed:
Merge branch 'main' into jwt

bradlarsen committed:
Merge branch 'main' into jwt

rosecodym commented on pkg/detectors/jwt/jwt.go:
```diff
@@ -0,0 +1,357 @@
+// Attempt to verify a JWT that uses an HMAC algorithm.
+//
+// This implementation only attempts to verify JWTs whose issuers use the OIDC Discovery protocol to make public keys available via request.
+func verifyHMAC(parsedToken *jwt.Token) (bool, error) {
```
I'm generally opposed to functions that signify they can return things they can't actually return (an `error` in this case), because I think it usually adds unnecessary cognitive overhead during future maintenance. (I'm not requesting a change - I'm just flagging this because it looks like the elimination of any actually returned error might have happened in the course of writing this function rather than as your initial plan.)

bradlarsen replied:
I've changed this in the latest version. HMAC-based JWTs are skipped entirely now, even before verification.

rosecodym reviewed (commented):
I have a question about some verification errors this returns.

bradlarsen committed:
Simplify JWT verification

- Be more careful with error return values to produce fewer `unknown`-status findings
- Combine the HMAC and public-key verifier implementation
- Perform JWT validation and verification explicitly instead of using `jwt.Parse`

bradlarsen committed:
Merge remote-tracking branch 'upstream' into jwt

bradlarsen committed:
Whoops: get rid of debugging code

bradlarsen committed:
Fix JWT signature verification; fix tests; clean up detector code

bradlarsen committed:
Merge remote-tracking branch 'upstream' into jwt

bradlarsen committed:
Merge branch 'main' into jwt
